### ISSUE TITLE
OnBefore hooks do not prevent the URL changing when returning a rejected promise

### ISSUE THREAD

**Issue by EmandM** [angular-ui/ui-router#2805]
When trying to implement an `onBefore` hook to prevent navigation until the user confirms, there is some unexpected behaviour. The docs contain conflicting information about this behaviour

Example hook

```
$transitions.onBefore(true, function () {
    return showModal();
});
```

When `showModal` returns a boolean the onBefore hook behaves as expected and the url does not change unless `showModal` returns true. However when `showModal` returns a promise the url updates to the new state whether or not `showModal` resolves to true.

Turning to [the docs](http://angular-ui.github.io/ui-router/feature-1.0/interfaces/transition.ihookregistry.html#onbefore) to check the expected behaviour, and the first thing you see is

> The Transition waits for this promise to settle before entering the onStart phase

but then in the next paragraph

> If any hook returns a promise, the remainder of the onBefore hooks are still invoked. Any promises are then handled asynchronously, during the onStart phase of the Transition.

It seems like `onBefore` handles promises in the onStart phase which does not make much sense in the sense of "do this behaviour before the transition starts". Can we please update this behaviour so that onBefore resolves promises before the onStart phase begins?

**Comment by christopherthielen** [comment]
I realize this is now three months later. I believe the error is in the wording of the documentation, not in the implementation.

The onStart phase Ddoes not begin until onBefore has completed, including all of its promises resolving

**Comment by christopherthielen** [comment]
It looks like this documentation has already been updated

> If any hook modifies the transition synchronously (by throwing, returning false, or returning a TargetState), the remainder of the hooks are skipped. If a hook returns a promise, the remainder of the onBefore hooks are still invoked synchronously. All promises are resolved, and processed asynchronously before the onStart phase of the Transition.

Closed by ca45957901d53a77f69074d15ea87baf1b78a48c

**Comment by sathishc** [comment]
I see the browser url changes even if I return false in onBefore hooks. using 1.0.5.  How can I make sure the url does not change?
